test/doc/developer/reverse.cpp

```diff
@@ -0,0 +1,194 @@
+    eve::aligned_ptr<T, size * sizeof(T)> buffer_data(buffer.data());
+
+    eve::store(x, buffer_data);
+    std::reverse(buffer.begin(), buffer.end());
+
+    return eve::wide<T, N, ABI>{buffer_data};
+  }
+  else if constexpr (std::same_as<ABI, eve::x86_128_>) return {_mm_shuffle_epi8(raw, reverse_16_chars_mask<N>())};
+  else if constexpr (std::same_as<ABI, eve::x86_256_>) return {reverse_256_using_128_mask(raw, reverse_16_chars_mask<N>())};
+}
+
+template <typename N>
+inline __m128i reverse_8_shorts_mask() {
+  return eve::wide<short, eve::fixed<8>>([](int i, int) {
+    auto as_chars_idx = [](int i) { return (i * 2 + 1) << 8 | (i * 2); };
+    return as_chars_idx(i >= N() ? i : (N() - i - 1));
+  }).storage();
+}
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> actual_reverse(eve::wide<T, N, ABI> x) requires(sizeof(T) == 2)
+{
+  auto raw = x.storage();
+
+       if constexpr(N() < 8) return {_mm_shufflelo_epi16(raw, mm_shuffle<N>())};
+  else if constexpr (eve::current_api <= eve::sse3)  // N == 8
+  {
+     raw = _mm_shufflehi_epi16(raw, _MM_SHUFFLE(0, 1, 2, 3));
+     raw = _mm_shufflelo_epi16(raw, _MM_SHUFFLE(0, 1, 2, 3));
+     raw = _mm_shuffle_epi32(raw, _MM_SHUFFLE(1, 0, 3, 2));
+
+     return {raw};
+  }
+  else if constexpr (std::same_as<ABI, eve::x86_128_>) return {_mm_shuffle_epi8(raw, reverse_8_shorts_mask<N>())};
+  else if constexpr (std::same_as<ABI, eve::x86_256_>) return {reverse_256_using_128_mask(raw, reverse_8_shorts_mask<N>())};
+}
+
+template <typename Wide>
+constexpr bool treat_like_aggregated() {
+  using ABI = typename Wide::abi_type;
+
+       if constexpr (eve::has_aggregated_abi_v<Wide>)        return true;
+  else if constexpr (eve::current_api == eve::avx &&
+                     std::same_as<ABI, eve::x86_256_> &&
+                     sizeof(eve::element_type_t<Wide>) <= 4) return true;
+  else                                                       return false;
+}
+
+}  // namespace _reverse
+
+template <typename T, typename N, typename ABI>
+eve::wide<T, N, ABI> reverse(eve::wide<T, N, ABI> x) {
+  using Wide = eve::wide<T, N, ABI>;
+
+  if constexpr (Wide::static_size == 1)
+    return x;
+  else if constexpr (_reverse::treat_like_aggregated<Wide>())
+  {
+    auto [l, h] = x.slice();
+    l = eve_extra::reverse(l);
+    h = eve_extra::reverse(h);
+    return Wide(h, l);
+  }
+  else if constexpr (std::same_as<T, float>)
+  {
+    auto as_ints = eve::bit_cast(x, eve::as_<eve::wide<std::int32_t, N>>{});
+    as_ints = eve_extra::reverse(as_ints);
+    return eve::bit_cast(as_ints, eve::as_<Wide>{});
+  }
+  else if constexpr (std::same_as<T, double>)
+  {
+    auto as_ints = eve::bit_cast(x, eve::as_<eve::wide<std::int64_t, N>>{});
+    as_ints = eve_extra::reverse(as_ints);
+    return eve::bit_cast(as_ints, eve::as_<Wide>{});
+  }
+  else return _reverse::actual_reverse(x);
+}
+
+}  // namespace eve_extra
+
+
+template<typename T, typename Test>
```

Comment:
Would you like if I pass over this section so you can use TTS instead of manual tests ?

Reply:
I would appreciate it.